crates/sui-transaction-builder/src/error.rs

```diff
@@ -0,0 +1,194 @@
+pub enum MissingDataError {
+    /// Missing object id
+    ObjectId,
+    /// Missing version for the respective object
+    Version(ObjectId),
+    /// Missing digest for the respective object
+    Digest(ObjectId),
+    /// Missing the sender for this transaction. A sender is always required when building a
+    /// transaction.
+    Sender,
+    /// Missing the gas objects for this transaction
+    GasObjects,
+    /// Missing the gas budget for this transaction
+    GasBudget,
+    /// Missing the gas price for this transaction
+    GasPrice,
+    /// Missing the object kind for this object
+    ObjectKind(ObjectId),
+    /// Missing the initial shared version for this object
+    InitialSharedVersion(ObjectId),
+    /// Missing the value data for this input that is of type Pure
+    PureValue,
+    /// Cannot determine the shared object mutability for this object
+    SharedObjectMutability(ObjectId),
+}
+
+/// Literals are not supported yet.
+#[derive(Debug, Clone)]
+pub struct UnsupportedLiteralError;
+
+/// General error type for the transaction builder.
+#[derive(Debug)]
+pub struct Error {
+    inner: Box<BuilderError>,
+}
+
+/// Specific error type for the transaction builder.
+#[derive(Debug)]
+pub struct BuilderError {
+    missing_data: bool,
+    conversion_error: bool,
+    source: Option<BoxError>,
+}
+
+impl Error {
+    /// Create an empty error with no source.
+    fn empty() -> Self {
+        Self {
+            inner: Box::new(BuilderError {
+                missing_data: false,
+                conversion_error: false,
+                source: None,
+            }),
+        }
+    }
+
+    /// Convert the given error into a generic error.
+    fn from_error<E: Into<BoxError>>(error: E) -> Self {
+        let boxed_error = error.into();
+        let result = Self::empty();
+
+        if let Some(conversion_error) = boxed_error.downcast_ref::<ConversionError>() {
+            result.with_conversion_error(conversion_error.clone())
+        } else if let Some(missing_data_error) = boxed_error.downcast_ref::<MissingDataError>() {
+            result.with_missing_data(missing_data_error.clone())
+        } else {
+            result.with_error(boxed_error)
+        }
+    }
+
+    /// Set the source to the given error.
+    fn with_error(mut self, error: BoxError) -> Self {
+        self.inner.source.replace(error);
+        self
+    }
+
+    /// Set the missing data variable to true and set the source to the given error.
+    fn with_missing_data(mut self, error: MissingDataError) -> Self {
+        self.inner.missing_data = true;
+        self.inner.source.replace(Box::new(error));
+        self
+    }
+
+    /// Set the conversion error to true and set the source to the given error.
+    fn with_conversion_error(mut self, error: ConversionError) -> Self {
+        self.inner.conversion_error = true;
+        self.inner.source.replace(Box::new(error));
+        self
+    }
+}
+
+impl std::fmt::Display for Error {
```

Comment:
I guess it's a matter of style. I reworked this, but I think for `Error` we should keep it as we need to have our own `impl std::error::Error for Error {fn source(...)}`

Reply:
I agree with @amnn here in that, we're already using thiserror for defining a few error enums so lets just continue with that to reduce the code here. I'd probably even be fine condensing the 3 different error types/enums into a single one just so that there's a single "error type" exported from this crate

Reply:
I reworked a bit the errors but I am not 100% satisfied with it.

Reply:
@bmwill do we want to keep the pattern of having a general `Error` type that wraps an Error in a Box?